testing.rs

```diff
@@ -0,0 +1,629 @@
+/// Inserts a suite of ink! unit tests intended for a contract implementing PSP22 trait.
+/// `$contract` argument should be the name of the contract struct.
+/// `$constructor` argument should be the name of a function, which initializes `$contract`
+/// with the given total supply of tokens.
+/// This macro should be invoked inside `#[ink::contract]` module.
+#[macro_export]
+macro_rules! tests {
+    ($contract:ident, $constructor:expr) => {
+        mod psp22_unit_tests {
+            use super::super::*;
+            use ink::env::{test::*, DefaultEnvironment as E};
+
+            type Event = <$contract as ink::reflect::ContractEventBase>::Type;
+
+            // Gathers all emitted events, skip `shift` first, decode the rest and return as vector
+            fn decode_events(shift: usize) -> Vec<Event> {
+                recorded_events()
+                    .skip(shift)
+                    .map(|e| <Event as scale::Decode>::decode(&mut &e.data[..]).unwrap())
+                    .collect()
+            }
+
+            // Asserts if the given event is a Transfer with particular from_, to_ and value_
+            fn assert_transfer(event: &Event, from_: AccountId, to_: AccountId, value_: u128) {
+                if let Event::Transfer(Transfer { from, to, value }) = event {
+                    assert_eq!(*from, Some(from_), "Transfer event: 'from' mismatch");
+                    assert_eq!(*to, Some(to_), "Transfer event: 'to' mismatch");
+                    assert_eq!(*value, value_, "Transfer event: 'value' mismatch");
+                } else {
+                    panic!("Event is not Transfer")
+                }
+            }
+
+            // Asserts if the given event is a Approval with particular owner_, spender_ and amount_
+            fn assert_approval(
+                event: &Event,
+                owner_: AccountId,
+                spender_: AccountId,
+                amount_: u128,
+            ) {
+                if let Event::Approval(Approval {
+                    owner,
+                    spender,
+                    amount,
+                }) = event
+                {
+                    assert_eq!(*owner, owner_, "Approval event: 'owner' mismatch");
+                    assert_eq!(*spender, spender_, "Approval event: 'spender' mismatch");
+                    assert_eq!(*amount, amount_, "Approval event: 'amount' mismatch");
+                } else {
+                    panic!("Event is not Approval")
+                }
+            }
+
+            #[ink::test]
+            fn constructor_works() {
+                let acc = default_accounts::<E>();
+                set_caller::<E>(acc.alice);
+                let supply = 1000;
+                let token = $constructor(supply);
```

Comment:
I'd let the user choose `caller` as well. Otherwise it's impossible to recover that information w/o reading this code